named tensor max pooling support

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31669 named tensor max pooling support**

Differential Revision: [D19240348](https://our.internmc.facebook.com/intern/diff/D19240348)

Commit:
named tensor max pooling support

[ghstack-poisoned]

zou3519 commented on test/test_namedtensor.py:
```diff
@@ -272,6 +272,19 @@ def test_diagonal(self):
         self.assertEqual(named_tensor.diagonal(outdim='E', dim1='B', dim2='D').names,
                          ['A', 'C', 'E'])
 
+    def test_max_polling(self):
```
We should test CUDA as well because the code path is different for CPU and CUDA.

zou3519 commented on test/test_namedtensor.py:
```diff
@@ -272,6 +272,19 @@ def test_diagonal(self):
+        self.assertEqual(named_tensor_2d.names, F.max_pool2d(named_tensor_2d, [2, 2]).names)
+        self.assertEqual(named_tensor_3d.names, F.max_pool3d(named_tensor_3d, [2, 2, 2]).names)
+
+        self.assertEqual(named_tensor_1d.names, F.max_pool1d_with_indices(named_tensor_1d, 2)[0].names)
```
The indices should have names as well. Because the indices have the same shape as the result, the indices should probably have the same names as the result.

i.e., if our input has names `list("NCHW")`, then the output has names `list("NCHW")` and the indices should also have names `list("NCHW")` because the data has the same dimension ordering.

Action item: we should propagate names to the indices and test that they have the same names.

Commit:
Update on "named tensor max pooling support"


Differential Revision: [D19240348](https://our.internmc.facebook.com/intern/diff/D19240348)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f270cbc5:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f270cbc59bc38f02cd59df1e4c32e3f2421126e7).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4108168) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=79534074) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=79534074))
<details>
<summary>
<code>Jan 02 22:41:11 caused by: Connection refused (os error 111)</code>
</summary>

```
Jan 02 22:41:11 +++ eval 'extract_trap_cmd ' 
Jan 02 22:41:11 ++++ extract_trap_cmd 
Jan 02 22:41:11 ++++ printf '%s\n' '' 
Jan 02 22:41:11 +++ printf '%s\n' cleanup 
Jan 02 22:41:11 ++ trap -- ' 
Jan 02 22:41:11 cleanup' EXIT 
Jan 02 22:41:11 ++ which sccache 
Jan 02 22:41:11 ++ sccache --stop-server 
Jan 02 22:41:11 Stopping sccache server... 
Jan 02 22:41:11 error: couldn't connect to server 
Jan 02 22:41:11 caused by: Connection refused (os error 111) 
Jan 02 22:41:11 ++ true 
Jan 02 22:41:11 ++ rm /var/lib/jenkins/sccache_error.log 
Jan 02 22:41:11 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Jan 02 22:41:11 ++ true 
Jan 02 22:41:11 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Jan 02 22:41:11 ++ SCCACHE_IDLE_TIMEOUT=1200 
Jan 02 22:41:11 ++ RUST_LOG=sccache::server=error 
Jan 02 22:41:11 ++ sccache --start-server 
Jan 02 22:41:11 Starting sccache server... 
Jan 02 22:41:11 ++ sccache --zero-stats 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zou3519 reviewed (approved):
lgtm, but let's try to clean up the testing before landing

Commit:
Update on "named tensor max pooling support"


Differential Revision: [D19240348](https://our.internmc.facebook.com/intern/diff/D19240348)

[ghstack-poisoned]

zou3519 commented on test/test_namedtensor.py:
```diff
@@ -272,6 +272,23 @@ def test_diagonal(self):
         self.assertEqual(named_tensor.diagonal(outdim='E', dim1='B', dim2='D').names,
                          ['A', 'C', 'E'])
 
+    def test_max_polling(self):
```
s/max_polling/max_pooling/

glaringlee commented on test/test_namedtensor.py:
```diff
@@ -272,6 +272,39 @@ def test_diagonal(self):
+        self.assertEqual(named_tensor_2d.names, F.max_pool2d_with_indices(named_tensor_2d, [2, 2])[0].names)
+        self.assertEqual(named_tensor_2d.names, F.max_pool2d_with_indices(named_tensor_2d, [2, 2])[1].names)
+        self.assertEqual(named_tensor_3d.names, F.max_pool3d_with_indices(named_tensor_3d, [2, 2, 2])[0].names)
+        self.assertEqual(named_tensor_3d.names, F.max_pool3d_with_indices(named_tensor_3d, [2, 2, 2])[1].names)
```
updated

glaringlee commented on test/test_namedtensor.py:
```diff
@@ -272,6 +272,39 @@ def test_diagonal(self):
+        self.assertEqual(named_tensor_3d.names, F.max_pool3d_with_indices(named_tensor_3d, [2, 2, 2])[1].names)
+
+    @unittest.skipIf(not TEST_CUDA, 'no CUDA')
+    def test_max_polling_gpu(self):
```
updated

Commit:
Update on "named tensor max pooling support"


Differential Revision: [D19240348](https://our.internmc.facebook.com/intern/diff/D19240348)

[ghstack-poisoned]

zou3519 commented on test/test_namedtensor.py:
```diff
@@ -272,6 +272,23 @@ def test_diagonal(self):
+            self.assertEqual(F.max_pool3d(named_tensor_3d, [2, 2, 2]).names, named_tensor_3d.names)
+
+            self.assertEqual(F.max_pool1d_with_indices(named_tensor_1d, 2)[0].names, named_tensor_1d.names)
+            self.assertEqual(F.max_pool1d_with_indices(named_tensor_1d, 2)[1].names, named_tensor_1d.names)
```
nit: (this isn't blocking) `F.max_pool1d_with_indices(named_tensor_1d, 2)` returns a tuple. We end up computing this twice. Instead, we can compute it once and then check the names on the result:
```
def check_tuple_return(op, inputs, expected_names):
    values, indices = op(*inputs)
    self.assertEqual(values.names, expected_names)
    self.assertEqual(indices.names, expected_names)

check_operator(F.max_pool1d_with_indices, [namedtensor_1d, 2], named_tensor_1d.names)
check_operator(F.max_pool2d_with_indices, [namedtensor_2d, [2, 2]], named_tensor_2d.names)
check_operator(F.max_pool3d_with_indices, [namedtensor_3d, [2, 2, 2]], named_tensor_3d.names)
```

This would improve how fast this test runs by doing less. It shouldn't really matter though so feel free to ship without this.

glaringlee replied:
Updated

Commit:
Update on "named tensor max pooling support"


Differential Revision: [D19240348](https://our.internmc.facebook.com/intern/diff/D19240348)

[ghstack-poisoned]
